design/20200305-spec-v1alpha2-dependency-parampassing.md

```diff
@@ -0,0 +1,96 @@
+# Component and Trait Dependency with Parameter Passing
+
+* Author: @wonderflow @hongchaodeng
+
+## Background
+
+In https://github.com/oam-dev/spec/issues/171, we have discussed our dependency stories and started to explore solutions.
+Now that we have designed and implemented mechanism to satisfy dependency requirements, we are proposing to add them into OAM spec.
+
+## Use Cases
+
+### Case 1: MySQL and Web
+
+Two components, MySQL and Web exists in the same AppConfig and Web depends on MySQL that:
+
+1. MySQL needs to start first;
+2. MySQL connection info needs to be injected into Web's environment variables or files;
+
+### Case 2: NAS FileSystem and MountTarget
+
+Two components, NAS_FileSystem and NAS_MountTarget exists in the same AppConfig and NAS_MountTarget depends on NAS_FileSystem that:
+
+1. NAS_FileSystem needs to be created first;
+2. NAS_MountTarget needs filesystemID from NAS_FileSystem to fill in NAS_MountTarget's spec.
+
+## Proposal
+
+### Component and Trait Dependency
+
+We are proposing to add `dependencies` in AppConfig as follows:
+
+```yaml
+kind: ApplicationConfiguration
+spec:
+- components:
+  - componentName: mysql
+  - componentName: web
+    dependencies:
+      - mysql
+```
+
+Here a component can describe its dependencies, or dependent components.
+
+For Case 1, we also suggest to use ServiceBinding in [trait-injector](https://github.com/oam-dev/trait-injector) to inject secret. It is a trait that should be applied before component starts. We propose to add `ordering` to express trait's ordering w.r.t the component.
+
+```yaml
+- componentName: web
+  dependencies:
+    - mysql
+  traits:
+    - ordering: preStart # applied before component starts
```

Comment:
Having dependency information makes sense. On the other hand, specifying at which stage something should be applied is imperative and not declarative. IMO, OAM should keep a declarative config as much as possible.

Reply:
Could you provide a suggestion for declarative approach?

Reply:
I was thinking on a way to infer when applying the config that it needs to run in preStart phase. Is there a scenario binding will happen in another phase? If not, we can simply automatically apply bindings on preStart - always. Then, the `ordering` attribute can be removed from the spec.